hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/fs/CachingSpaceUsageSource.java

```diff
@@ -0,0 +1,141 @@
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdds.fs;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.util.OptionalLong;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * Stores space usage and refreshes it periodically.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+public class CachingSpaceUsageSource implements SpaceUsageSource {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(CachingSpaceUsageSource.class);
+
+  private final ScheduledExecutorService executor;
+  private final AtomicLong cachedValue = new AtomicLong();
+  private final Duration refresh;
+  private final SpaceUsageSource source;
+  private final SpaceUsagePersistence persistence;
+  private boolean running;
+  private ScheduledFuture<?> scheduledFuture;
+
+  public CachingSpaceUsageSource(SpaceUsageCheckParams params) {
+    this(params, createExecutor(params));
+  }
+
+  public CachingSpaceUsageSource(SpaceUsageCheckParams params,
+      ScheduledExecutorService executor) {
+    Preconditions.checkArgument(params != null, "params == null");
+
+    refresh = params.getRefresh();
+    source = params.getSource();
+    persistence = params.getPersistence();
+    this.executor = refresh.isZero() ? null : executor;
+
+    loadInitialValue();
+  }
+
+  @Override
+  public long getCapacity() {
+    return source.getCapacity();
+  }
+
+  @Override
+  public long getAvailable() {
+    return source.getAvailable();
+  }
+
+  @Override
+  public long getUsedSpace() {
+    return cachedValue.get();
+  }
+
+  public void start() {
+    if (executor != null) {
+      long initialDelay = cachedValue.get() > 0 ? refresh.toMillis() : 0;
+      synchronized (this) {
+        if (!running) {
+          scheduledFuture = executor.scheduleWithFixedDelay(
+              this::refresh, initialDelay, refresh.toMillis(), MILLISECONDS);
+          running = true;
+        }
+      }
+    } else {
+      refresh();
+    }
+  }
+
+  public void shutdown() {
+    synchronized (this) {
```

Comment:
I don't think we need the synchronization. Looks like we only synchronize in stop and start, and user of the class should ensure they don't invoke both concurrently.

Reply:
Good point, `VolumeUsage` already synchronizes start and shutdown.